### PR TITLE
[X-Ref: #179] Begin adding tests for CLI

### PR DESCRIPTION
**Pull request by kkirsche**
[X-Ref: #179] Begin adding tests for CLI

### CODE REVIEW

**Comment by cvrebert** [comment]
Please don't include the `dist` files; they make doing the review a bit more annoying.

File: test/cli_test.js [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+        test.doesNotThrow(block, [error], [message])
+        test.ifError(value)
+*/
+
+exports.bootlint = {
+    setUp: function (done) {
+        // setup here
+        done();
+    },
+    'Program Details': function (test) {
```
Comment (cvrebert) [inline test/cli_test.js]: For consistency with the other tests, don't use Title Case for these keys

Comment (kkirsche) [inline test/cli_test.js]: Changed.

**Comment by kkirsche** [comment]
@cvrebert `dist` file has been removed from the PR. Hopefully this is at least a good starting point. Not 100% sure how best to test this but figured something is better than nothing as a starting point.

File: src/cli.js [hunk 2]
```diff
@@ -1,105 +1,115 @@
...
-
-function buildReporter(origin) {
-    return function (lint) {
-        var lintId = (lint.id[0] === 'E') ? chalk.bgGreen.white(lint.id) : chalk.bgRed.white(lint.id);
-        var output = false;
-        if (lint.elements) {
-            lint.elements.each(function (_, element) {
-                var loc = element.startLocation;
-                console.log(origin + ":" + (loc.line + 1) + ":" + (loc.column + 1), lintId, lint.message);
+(function () {
```
Comment (hnrch02) [inline src/cli.js]: Why is this IIFE needed?

Comment (kkirsche) [inline src/cli.js]: Removed IIFE.

**Comment by kkirsche** [comment]
Let me know if there is anything that I can do to improve these or get more coverage. I appreciate the time, reviews and advice. I know you guys are busy :+1:

File: src/cli.js [hunk 3]
```diff
@@ -43,6 +43,8 @@ function buildReporter(origin) {
     };
 }
 
+exports.buildReporter = buildReporter;
```
Comment (hnrch02) [inline src/cli.js]: I'd rather see us using https://github.com/jhnns/rewire for this.

File: test/cli_test.js [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+        test.strictEqual(cli.program.option('').options[1].optional, 0, 'should return whether the version flag is optional.');
+        test.strictEqual(cli.program.option('').options[1].bool, true, 'should return whether the version flag is optional.');
+        test.strictEqual(cli.program.option('').options[1].short, '-d', 'should list the short disable flag');
+        test.strictEqual(cli.program.option('').options[1].long, '--disable', 'should list the long disable flag');
+        test.strictEqual(cli.program.option('').options[1].description, 'Comma-separated list of disabled lint problem IDs',
+            'should provide a description of the disable flag');
+        test.done();
+    },
+    'Handle standard input': function (test) {
+        test.expect(6);
```
Comment (hnrch02) [inline test/cli_test.js]: Perhaps you could work with https://github.com/cjohansen/Sinon.JS and actually stub out `process.stdin`.

**Comment by hnrch02** [comment]
Referencing JSCS' CLI tests here: https://github.com/jscs-dev/node-jscs/blob/master/test/cli.js

File: src/cli.js (hunk 3 above)
Comment (kkirsche) [inline src/cli.js]: @hnrch02 Sorry, I'm sure I'm missing something simple here, and I don't want to be a nuisance but I'm confused about how rewire exposes the program. I don't mind using rewire but how do I expose the "program" to the tests if I don't export anything? I'll concede I'm exposing the wrong thing here and I need to expose the primary "CLI tool" but not sure how to do that here. Rewire as it says works like require, but with the ability to mock stuff. So do I need to wrap the CLI in `module.exports` like node-jscs like they do in `cli.js` or something else.

Sorry again.

**Comment by kkirsche** [comment]
Build failed due to: `The command "./test-infra/s3_cache.py download rubygems" failed and exited with 1 during .`

**Comment by cvrebert** [comment]
Looks like RubyGems was flaky:
`bad response Bad Gateway 502 (https://rubygems.global.ssl.fastly.net/gems/coffee-script-source-1.9.1.gem)`

*mashes Restart Build button*

**Comment by cvrebert** [comment]
@hnrch02 LGTY?

**Comment by hnrch02** [comment]
:+1:

**Pull request merged**